### PR TITLE
Implement eth_getBlockByNumber while downloading state.

### LINKED ISSUE #2209: Determine correct behavior of eth_getBlockByNumber, latest
Currently, we return the latest chain head even if we don't have state for it. For example, in fast sync. At least one other client seems to return the latest head that we have the state for, which means while fast syncing they return 0. We should work out what the behavior should be.

This has implications for how hive testing is currently done.

### PR DESCRIPTION
**Pull request by jflo**
## PR description
Introduces a dependency from eth_getBlockByNumber onto the Synchronizer, in order to determine download state, and current most recent block that we have state for. JsonRPC test classes also were refactored to provide a Synchronizer.

## Fixed Issue(s)

fixes #2209 

## Changelog
When asking for the latest block while downloading state, will return most recent block we have state for. While still downloading block headers, will return genesis block.

- [X] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

**Comment by vmichalik** [comment]
Hey @jflo, it would be useful for posterity to have a descriptive PR title. How about "Implement correct behavior of eth_getBlockByNumber"?

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetBlockByNumber.java [hunk 1]
```diff
@@ -20,27 +20,39 @@
...
 
 import com.google.common.base.Suppliers;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 
 public class EthGetBlockByNumber extends AbstractBlockParameterMethod {
 
   private final BlockResultFactory blockResult;
   private final boolean includeCoinbase;
+  private static final Logger log = LogManager.getLogger();
```
Comment (matkt) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetBlockByNumber.java:38]: For more consistency with the other classes I would say that it is necessary to do like this
> private static final Logger LOGGER = LogManager.getLogger();

File: ethereum/retesteth/src/main/java/org/hyperledger/besu/ethereum/retesteth/DummySynchronizer.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+package org.hyperledger.besu.ethereum.retesteth;
+
+import org.hyperledger.besu.ethereum.core.Synchronizer;
+import org.hyperledger.besu.plugin.data.SyncStatus;
+import org.hyperledger.besu.plugin.services.BesuEvents;
+
+import java.util.Optional;
+
+public class DummySynchronizer implements Synchronizer {
```
Comment (matkt) [inline ethereum/retesteth/src/main/java/org/hyperledger/besu/ethereum/retesteth/DummySynchronizer.java:24]: NIT : I would have added a description to the class to understand the goal of this class faster

**Review by matkt — APPROVED** [review]
LGTM

**Pull request merged**